### PR TITLE
Use .NET 5 docker image

### PR DESCRIPTION
**Pull request by russcam**
## What does this PR do?

This PR updates the dotnet Dockerfile to use .NET 5 docker image for build.

`DOTNET_HOST_PATH` needs to be set to workaround an issue with building ASP.NET Core 2.1 with the .NET 5 SDKs. See https://github.com/dotnet/sdk/issues/14497 for more details.

## Why is it important?

Needed to be able to run integration tests for the .NET agent when the global SDK is updated to .NET 5.

## Related issues
elastic/apm-agent-dotnet#1040

### CODE REVIEW

**Comment by v1v** [comment]
I've just triggered the `.NET` apm-its -> [build](https://apm-ci.elastic.co/job/apm-integration-test-downstream/job/PR-995/2/) since the issue is with the java agent when fetching artifacts from the maven central, we are working on providing a more stable build by using the cached artifacts

**Comment by russcam** [comment]
@v1v do I need to do anything further for this? Looking at the output, it looks like the issue is with the Java agent still?

**Comment by v1v** [comment]
Hi @russcam,

As far as I see the latest issue was not related to java but the agent-donet itself:

```
[2020-12-08T04:18:30.068Z] Service 'agent-dotnet' failed to build : The command '/bin/sh -c ./run.sh' returned a non-zero code: 145
```

See [logs](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-integration-test-downstream/detail/PR-995/13/pipeline#step-83-log-298)

You can see the docker log details in:
- https://apm-ci.elastic.co/job/apm-integration-test-downstream/job/PR-995/13/artifact/docker-info/all/

As far as I see the importan one is [0c15c67b1249.log](https://apm-ci.elastic.co/job/apm-integration-test-downstream/job/PR-995/13/artifact/docker-info/all/0c15c67b1249.log/*view*/) that shows an error like

```
Error response from daemon: configured logging driver does not support reading
```

You can reproduce it locally either use the `all.sh` wrapper or the `scripts/compose.py`:

### All.sh

```bash
ELASTIC_STACK_VERSION=8.0.0 .ci/scripts/all.sh
```

### Compose.py 

```bash
python3 scripts/compose.py start 8.0.0  \
 --build-parallel \
 --java-m2-cache \
 --no-apm-server-dashboards  \
 --no-apm-server-self-instrument   \
 --with-agent-rumjs  \
 --with-agent-dotnet   \
 --with-agent-go-net-http  \
 --with-agent-nodejs-express  \
 --with-agent-php-apache  \
 --with-agent-ruby-rails   \
 --with-agent-java-spring   \
 --with-agent-python-django  \
 --with-agent-python-flask   \
 --force-build   \
 --no-xpack-secure   \
 --apm-server-enable-tls   \
 --no-verify-server-cert    \
 --apm-server-secret-token=SuPeRsEcReT   \
 --apm-server-url=https://apm-server:8200  \
 --apm-log-level=debug
```

**Comment by v1v** [comment]
The below script should help you to just build the agent dotnet:

```bash
python3 scripts/compose.py start 8.0.0  \
 --no-apm-server-dashboards  \
 --no-apm-server-self-instrument   \
 --with-agent-dotnet   \
 --force-build   \
 --no-xpack-secure   \
 --apm-server-enable-tls   \
 --no-verify-server-cert    \
 --apm-server-secret-token=SuPeRsEcReT   \
 --apm-server-url=https://apm-server:8200  \
 --apm-log-level=debug
```

```
Your branch is up to date with 'origin/master'.
+ git rev-parse HEAD
8875c8a2fdd514d3ece3151742d03816daa9fa08
+ mkdir /src/local-packages
+ set -e
+ '[' -e .ci/linux/remove-projects.sh ']'
+ .ci/linux/remove-projects.sh
+ dotnet sln remove sample/AspNetFullFrameworkSampleApp/AspNetFullFrameworkSampleApp.csproj
Could not execute because the application was not found or a compatible .NET SDK is not installed.
Possible reasons for this include:
  * You intended to execute a .NET program:
      The application 'sln' does not exist.
  * You intended to execute a .NET SDK command:
      A compatible installed .NET SDK for global.json version [3.1.100] from [/src/dotnet-agent/global.json] was not found.
      Install the [3.1.100] .NET SDK or update [/src/dotnet-agent/global.json] with an installed .NET SDK:
        5.0.100 [/usr/share/dotnet/sdk]
ERROR: Service 'agent-dotnet' failed to build : The command '/bin/sh -c ./run.sh' returned a non-zero code: 145
Traceback (most recent call last):
  File "/Users/vmartinez/work/src/github.com/elastic/apm-integration-testing/scripts/compose.py", line 19, in <module>
    main()
  File "/Users/vmartinez/work/src/github.com/elastic/apm-integration-testing/scripts/compose.py", line 15, in main
    setup()
  File "/Users/vmartinez/work/src/github.com/elastic/apm-integration-testing/scripts/modules/cli.py", line 190, in __call__
    self.args.func()
  File "/Users/vmartinez/work/src/github.com/elastic/apm-integration-testing/scripts/modules/cli.py", line 660, in start_handler
    self.run_docker_compose_process(docker_compose_build + build_services)
  File "/Users/vmartinez/work/src/github.com/elastic/apm-integration-testing/scripts/modules/cli.py", line 439, in run_docker_compose_process
    subprocess.check_call(docker_compose_cmd)
  File "/usr/local/Cellar/python@3.9/3.9.0_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['docker-compose', '-f', '/Users/vmartinez/work/src/github.com/elastic/apm-integration-testing/docker-compose.yml', 'build', '--pull', '--no-cache', 'agent-dotnet']' returned non-zero exit status 1.
```

**Comment by russcam** [comment]
Thanks @v1v, this was super helpful! 

I've added a check to see if the SDK version specified in global.json is installed, and to install it if not. I think this is somewhat a chicken and egg problem for now, as we'll have some PRs with a 3.1 SDK in global.json, and others with a 5.0 SDK, until at least https://github.com/elastic/apm-agent-dotnet/pull/1040 is merged and PRs updated.

**Comment by russcam** [comment]
Error now is

```
[2020-12-09T06:27:55.983Z] ERROR: for agent-dotnet  (<Service: agent-dotnet>, "The command '/bin/sh -c ./run.sh' returned a non-zero code: 134")
[2020-12-09T06:27:55.983Z] Service 'agent-dotnet' failed to build : The command '/bin/sh -c ./run.sh' returned a non-zero code: 134
```

It's working as expected on my local machine

```
Step 9/15 : RUN ./run.sh
 ---> Running in d4a6510ad6f2
+ CSPROJ=TestAspNetCoreApp.csproj
+ PACKAGE=Elastic.Apm.NetCoreAll
+ CSPROJ_VERSION=/src/dotnet-agent/src/Elastic.Apm.NetCoreAll/Elastic.Apm.NetCoreAll.csproj
+ BUILD_PROPS=/src/dotnet-agent/src/Directory.Build.props
+ '[' -z '' ']'
+ git clone https://github.com/elastic/apm-agent-dotnet.git /src/dotnet-agent
Cloning into '/src/dotnet-agent'...
+ cd /src/dotnet-agent
+ git fetch -q origin '+refs/pull/*:refs/remotes/origin/pr/*'
error: cannot lock ref 'refs/remotes/origin/pr/82/head': 'refs/remotes/origin/pr/82' exists; cannot create 'refs/remotes/origin/pr/82/head'
error: cannot lock ref 'refs/remotes/origin/pr/855/head': 'refs/remotes/origin/pr/855' exists; cannot create 'refs/remotes/origin/pr/855/head'
+ git checkout master
Already on 'master'
Your branch is up to date with 'origin/master'.
+ git rev-parse HEAD
8875c8a2fdd514d3ece3151742d03816daa9fa08
+ [[ -f global.json ]]
++ sed -n -e 's/\s*"version":\s*"\(.*\)",/\1/p' global.json
+ GLOBAL_JSON_VERSION=3.1.100
+ INSTALLED_VERSIONS=()
+ IFS=
+ read -r result
++ dotnet --list-sdks
+ INSTALLED_VERSIONS+=("$(echo "$result" | sed -e 's/^\(.*\)\s.*/\1/')")
++ echo '5.0.101 [/usr/share/dotnet/sdk]'
++ sed -e 's/^\(.*\)\s.*/\1/'
+ IFS=
+ read -r result
+ [[ !  5.0.101  =~  3\.1\.100  ]]
+ curl -s -O -L https://dotnet.microsoft.com/download/dotnet-core/scripts/v1/dotnet-install.sh
+ bash ./dotnet-install.sh --install-dir /usr/share/dotnet -version 3.1.100
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/3.1.100/dotnet-sdk-3.1.100-linux-x64.tar.gz
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/3.1.100/dotnet-sdk-3.1.100-linux-x64.tar.gz
dotnet-install: Adding to current process PATH: `/usr/share/dotnet`. Note: This change will be visible only when sourcing script.
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install, select your operating system and check the "Dependencies" section.
dotnet-install: Installation finished successfully.
+ mkdir /src/local-packages
+ set -e
+ '[' -e .ci/linux/remove-projects.sh ']'
+ .ci/linux/remove-projects.sh
+ dotnet sln remove sample/AspNetFullFrameworkSampleApp/AspNetFullFrameworkSampleApp.csproj

Welcome to .NET Core 3.1!
---------------------
SDK Version: 3.1.100

Telemetry
---------
The .NET Core tools collect usage data in order to help us improve your experience. The data is anonymous. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET Core CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Find out what's new: https://aka.ms/dotnet-whats-new
Learn about the installed HTTPS developer cert: https://aka.ms/aspnet-core-https
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli-docs
Write your first app: https://aka.ms/first-net-core-app
--------------------------------------------------------------------------------------
Project `sample/AspNetFullFrameworkSampleApp/AspNetFullFrameworkSampleApp.csproj` removed from the solution.
+ dotnet sln remove src/Elastic.Apm.AspNetFullFramework/Elastic.Apm.AspNetFullFramework.csproj
Project `src/Elastic.Apm.AspNetFullFramework/Elastic.Apm.AspNetFullFramework.csproj` removed from the solution.
+ dotnet sln remove test/Elastic.Apm.AspNetFullFramework.Tests/Elastic.Apm.AspNetFullFramework.Tests.csproj
Project `test/Elastic.Apm.AspNetFullFramework.Tests/Elastic.Apm.AspNetFullFramework.Tests.csproj` removed from the solution.
+ dotnet restore

...
```

File: .ci/scripts/common.sh [hunk 1]
```diff
@@ -28,7 +28,7 @@ if [ -n "${APM_SERVER_BRANCH}" ]; then
 fi
 
 if [ -z "${DISABLE_BUILD_PARALLEL}" -o "${DISABLE_BUILD_PARALLEL}" = "false" ]; then
- BUILD_OPTS="${BUILD_OPTS} --build-parallel"
+ BUILD_OPTS="${BUILD_OPTS}"
```
Comment (v1v) [inline .ci/scripts/common.sh:31]:
```suggestion
 BUILD_OPTS="${BUILD_OPTS} --build-parallel"
```

File: docker/dotnet/Dockerfile [hunk 2]
```diff
@@ -3,12 +3,18 @@
...
+ENV DOTNET_ROOT=/usr/share/dotnet
 ARG DOTNET_AGENT_REPO=elastic/apm-agent-dotnet
 ARG DOTNET_AGENT_BRANCH=master
 ARG DOTNET_AGENT_VERSION=
+# Workaround for https://github.com/dotnet/sdk/issues/14497
+ARG DOTNET_HOST_PATH=/usr/share/dotnet/dotnet
 WORKDIR /src
 COPY . /src
+# install SDK version in global.json of elastic/apm-agent-dotnet
+# TODO: remove once global.json is on 5.0 *and* PRs are updated to 5.0 in global.json
```
Comment (v1v) [inline docker/dotnet/Dockerfile:16]: I guess for backward compatibility reasons we might need to keep it

```suggestion
# Backward compatibility when global.json is *not* on 5.0
```

Comment (russcam) [reply docker/dotnet/Dockerfile:16]: This is only needed whilst there are branches that build on CI that still specify `3.1.100` as the SDK to use in global.json. At some point in the near future after this PR is merged and global.json is updated to `5.0.100` SDK in https://github.com/elastic/apm-agent-dotnet/pull/1040, existing open PRs specifying `3.1.100` can be rebased to bring in the change to `5.0.100`, and new PRs will specify `5.0.100` as the SDK to use. From then, I don't think we need to install `3.1.100` SDK.

Comment (russcam) [reply docker/dotnet/Dockerfile]: I've updated the comments

**Comment by russcam** [comment]
Looks like the remaining failure is linting the shell scripts. 

@v1v does parallelizing the build need to be reinstated that was removed in https://github.com/elastic/apm-integration-testing/pull/995/commits/91f18e3bba8323d361b3af55af7e6d056f480fde?

**Comment by v1v** [comment]
I already reverted it with https://github.com/elastic/apm-integration-testing/commit/49b5f697829d34e7099ebd3f7725e638a7cc9d75

Is that what you meant?

**Comment by v1v** [comment]
jenkins run the tests please

**Review by v1v — APPROVED** [review]

**Comment by russcam** [comment]
> I already reverted it with 49b5f69
> Is that what you meant?

I missed that one! Yep ☺️

**Pull request merged**